File: system/cli/ops.py
"""brana ops — scheduler and system operations CLI."""

from __future__ import annotations

import json
import re
import subprocess
from datetime import datetime
from pathlib import Path
from typing import Optional

import typer

from .config import (
    SCHEDULER_CONFIG, SCHEDULER_LOGS, SCHEDULER_STATUS, SCHEDULER_TEMPLATE,
    detect_project, load_json,
)
from .theme import color, console, create_table, get_theme, icon, print_header

ops_app = typer.Typer()


# ── helpers ──────────────────────────────────────────────────────────────


def _load_scheduler() -> dict:
    data = load_json(SCHEDULER_CONFIG)
    return data if data else {"jobs": {}}


def _load_status() -> dict:
    data = load_json(SCHEDULER_STATUS)
    return data if data else {}


def _get_timer_info(job_name: str) -> dict:
    unit = f"brana-sched-{job_name}.timer"
    try:
        result = subprocess.run(
            ["systemctl", "--user", "show", unit,
             "--property=ActiveState,NextElapseUSecRealtime,Result"],
            capture_output=True, text=True, timeout=5,
        )
        info = {}
        for line in result.stdout.strip().splitlines():
            if "=" in line:
                k, v = line.split("=", 1)
                info[k] = v
        return info
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return {}


def _status_icon(status: str, theme: dict) -> str:
    mapping = {
        "SUCCESS": "done",
        "FAILED": "blocked",
        "TIMEOUT": "blocked",
        "SKIPPED": "parked",
    }
    return icon(mapping.get(status, "pending"), theme)


def _status_color(status: str) -> str:
    mapping = {
        "SUCCESS": "green",
        "FAILED": "red",
        "TIMEOUT": "red",
        "SKIPPED": "dim",
    }
    return mapping.get(status, "yellow")


# ── commands ─────────────────────────────────────────────────────────────


@ops_app.command()
def status(wide: bool = typer.Option(False, "--wide", "-w")):
    """Dashboard: all jobs, last run, next trigger, health."""
    theme = get_theme()
    sched = _load_scheduler()
    last_status = _load_status()
    jobs = sched.get("jobs", {})

    print_header("Scheduler", theme)

    if wide:
        table = create_table("", "Job", "Schedule", "Enabled", "Last Run", "Status",
                             "Next Trigger", "Project")
        for name, cfg in sorted(jobs.items()):
            enabled = cfg.get("enabled", True)
            job_status = last_status.get(name, {})
            st = job_status.get("status", "—")
            ts = job_status.get("timestamp", "—")
            if ts != "—":
                try:
                    ts = datetime.fromisoformat(ts).strftime("%Y-%m-%d %H:%M")
                except ValueError:
                    pass

            timer = _get_timer_info(name)
            next_trigger = timer.get("NextElapseUSecRealtime", "—")
            if next_trigger and next_trigger != "—":
                # systemd gives epoch microseconds or human-readable
                try:
                    next_trigger = next_trigger.split(" ")[0:3]
                    next_trigger = " ".join(next_trigger)
                except (IndexError, ValueError):
                    pass

            ic = _status_icon(st, theme) if enabled else icon("parked", theme)
            table.add_row(
                ic, name, cfg.get("schedule", "—"),
                "yes" if enabled else "no",
                ts, st, str(next_trigger),
                Path(cfg.get("project", "")).name,
                style=_status_color(st) if enabled else "dim",
            )
        console.print(table)
    else:
        for name, cfg in sorted(jobs.items()):
            enabled = cfg.get("enabled", True)
            job_status = last_status.get(name, {})
            st = job_status.get("status", "—")
            ts = job_status.get("timestamp", "")
            if ts:
                try:
                    ts = datetime.fromisoformat(ts).strftime("%m-%d %H:%M")
                except ValueError:
                    pass

            ic = _status_icon(st, theme) if enabled else icon("parked", theme)
            schedule = cfg.get("schedule", "—")
            style = _status_color(st) if enabled else "dim"

            line = f"  {ic} {name:<24} {schedule:<24} {st:<10} {ts}"
            if not enabled:
                line += "  [disabled]"
            console.print(line, style=style)

    console.print()


@ops_app.command()
def health():
    """Aggregate health: failures in 24h, missed runs, lock contention."""
    theme = get_theme()
    sched = _load_scheduler()
    last_status = _load_status()
    jobs = sched.get("jobs", {})

    now = datetime.now()
    failures_24h = []
    skipped_24h = []

    for name, info in last_status.items():
        ts = info.get("timestamp", "")
        try:
            run_time = datetime.fromisoformat(ts)
            if (now - run_time).total_seconds() > 86400:
                continue
        except ValueError:
            continue

        if info.get("status") == "FAILED":
            failures_24h.append(name)
        elif info.get("status") == "SKIPPED":
            skipped_24h.append(name)

    # Check for schedule collisions
    collisions = _find_collisions(jobs)

    print_header("Scheduler health", theme)

    ok = icon("done", theme)
    fail = icon("blocked", theme)
    warn = icon("pending", theme)

    if failures_24h:
        console.print(f"  {fail} Failures (24h): {', '.join(failures_24h)}", style="red")
    else:
        console.print(f"  {ok} No failures in 24h", style="green")

    if skipped_24h:
        console.print(f"  {warn} Skipped/locked: {', '.join(skipped_24h)}", style="yellow")

    if collisions:
        console.print(f"  {fail} Schedule collisions:", style="red")
        for group in collisions:
            console.print(f"      {group['schedule']} on {group['project']}: {', '.join(group['jobs'])}")
    else:
        console.print(f"  {ok} No schedule collisions", style="green")

    enabled = sum(1 for j in jobs.values() if j.get("enabled", True))
    disabled = len(jobs) - enabled
    console.print(f"\n  {enabled} enabled, {disabled} disabled, {len(jobs)} total")
    console.print()


def _find_collisions(jobs: dict) -> list[dict]:
    groups: dict[tuple, list] = {}
    for name, cfg in jobs.items():
        if not cfg.get("enabled", True):
            continue
        key = (cfg.get("schedule", ""), Path(cfg.get("project", "")).name)
        groups.setdefault(key, []).append(name)

    return [
        {"schedule": k[0], "project": k[1], "jobs": v}
        for k, v in groups.items()
        if len(v) > 1
    ]


@ops_app.command()
def logs(
    job_name: str = typer.Argument(..., help="Job name"),
    tail: int = typer.Option(50, "--tail", "-n", help="Number of lines"),
):
    """View logs for a scheduler job."""
    job_log_dir = SCHEDULER_LOGS / job_name
    if not job_log_dir.exists():
        console.print(f"\n  No logs found for '{job_name}'.\n")
        raise typer.Exit(1)

    log_files = sorted(job_log_dir.glob("*.log"), reverse=True)
    if not log_files:
        console.print(f"\n  No log files in {job_log_dir}.\n")
        raise typer.Exit(1)

    latest = log_files[0]
    lines = latest.read_text().splitlines()
    show = lines[-tail:] if len(lines) > tail else lines

    console.print(f"\n  [dim]{latest.name}[/]\n")
    for line in show:
        if "SUCCESS" in line:
            console.print(f"  [green]{line}[/]")
        elif "FAILED" in line or "ERROR" in line:
            console.print(f"  [red]{line}[/]")
        elif "TIMEOUT" in line or "SKIPPED" in line:
            console.print(f"  [yellow]{line}[/]")
        else:
            console.print(f"  {line}")
    console.print()


@ops_app.command()
def history(
    job_name: str = typer.Argument(..., help="Job name"),
    last: int = typer.Option(10, "--last", "-n", help="Number of runs to show"),
):
    """Run history for a job (pass/fail trend)."""
    theme = get_theme()
    job_log_dir = SCHEDULER_LOGS / job_name
    if not job_log_dir.exists():
        console.print(f"\n  No history for '{job_name}'.\n")
        raise typer.Exit(1)

    log_files = sorted(job_log_dir.glob("*.log"), reverse=True)[:last]

    print_header(f"History: {job_name} (last {len(log_files)})", theme)
    for lf in log_files:
        content = lf.read_text()
        if "SUCCESS" in content:
            st, style = "SUCCESS", "green"
        elif "FAILED" in content:
            st, style = "FAILED", "red"
        elif "TIMEOUT" in content:
            st, style = "TIMEOUT", "red"
        elif "SKIPPED" in content:
            st, style = "SKIPPED", "dim"
        else:
            st, style = "UNKNOWN", "yellow"

        date = lf.stem  # filename is timestamp
        console.print(f"  [{style}]{_status_icon(st, theme)} {date}  {st}[/]")

    console.print()


@ops_app.command()
def collisions():
    """Detect same-project schedule conflicts."""
    theme = get_theme()
    sched = _load_scheduler()
    groups = _find_collisions(sched.get("jobs", {}))

    if not groups:
        console.print(f"\n  {icon('done', theme)} No schedule collisions.\n", style="green")
        return

    print_header("Schedule collisions", theme)
    for g in groups:
        console.print(
            f"  {icon('blocked', theme)} {g['schedule']} on {g['project']}: "
            f"{', '.join(g['jobs'])}",
            style="red",
        )
    console.print()


@ops_app.command()
def drift():
    """Compare live scheduler config vs template."""
    theme = get_theme()

    if SCHEDULER_TEMPLATE is None or not SCHEDULER_TEMPLATE.exists():
        console.print("\n  Template not found.\n")
        raise typer.Exit(1)

    template = load_json(SCHEDULER_TEMPLATE)
    live = _load_scheduler()

    if template is None:
        console.print("\n  Cannot load template.\n")
        raise typer.Exit(1)

    tmpl_jobs = template.get("jobs", {})
    live_jobs = live.get("jobs", {})

    drifts = []

    for name in set(tmpl_jobs) | set(live_jobs):
        if name not in tmpl_jobs:
            drifts.append(f"  [yellow]+ {name}: in live but not in template[/]")
        elif name not in live_jobs:
            drifts.append(f"  [red]- {name}: in template but not in live[/]")
        else:
            for field in ["schedule", "enabled", "command", "project", "type"]:
                tv = tmpl_jobs[name].get(field)
                lv = live_jobs[name].get(field)
                if tv != lv:
                    drifts.append(f"  [yellow]~ {name}.{field}: template={tv} live={lv}[/]")

    if not drifts:
        console.print(f"\n  {icon('done', theme)} No drift — live matches template.\n", style="green")
        return

    print_header("Config drift (template vs live)", theme)
    for line in drifts:
        console.print(line)
    console.print()


_JOB_NAME_RE = re.compile(r"^[a-zA-Z0-9_-]+$")


def _validate_job_name(job_name: str):
    if not _JOB_NAME_RE.match(job_name):
        console.print(f"\n  Invalid job name '{job_name}'. Use alphanumeric, hyphens, underscores.\n")
        raise typer.Exit(1)


@ops_app.command()
def run(job_name: str = typer.Argument(..., help="Job to trigger manually")):
    """Manually trigger a scheduler job now."""
    _validate_job_name(job_name)
    sched = _load_scheduler()
    if job_name not in sched.get("jobs", {}):
        console.print(f"\n  Job '{job_name}' not found in scheduler config.\n")
        raise typer.Exit(1)

    unit = f"brana-sched-{job_name}.service"
    console.print(f"\n  Starting {unit}...")

    try:
        subprocess.run(
            ["systemctl", "--user", "start", unit],
            check=True, timeout=10,
        )
        console.print(f"  [green]Triggered. Check logs: brana ops logs {job_name}[/]\n")
    except subprocess.CalledProcessError as e:
        console.print(f"  [red]Failed to start: {e}[/]\n")
        raise typer.Exit(1)


@ops_app.command()
def enable(job_name: str = typer.Argument(..., help="Job to enable")):
    """Enable a disabled scheduler job."""
    _toggle_job(job_name, enabled=True)


@ops_app.command()
def disable(job_name: str = typer.Argument(..., help="Job to disable")):
    """Disable a scheduler job."""
    _toggle_job(job_name, enabled=False)


def _toggle_job(job_name: str, enabled: bool):
    _validate_job_name(job_name)
    sched = _load_scheduler()
    jobs = sched.get("jobs", {})

    if job_name not in jobs:
        console.print(f"\n  Job '{job_name}' not found.\n")
        raise typer.Exit(1)

    jobs[job_name]["enabled"] = enabled
    SCHEDULER_CONFIG.write_text(json.dumps(sched, indent=2) + "\n")

    action = "Enabled" if enabled else "Disabled"
    style = "green" if enabled else "yellow"
    console.print(f"\n  [{style}]{action} '{job_name}' in scheduler.json.[/]")

    # Also toggle systemd timer
    timer_unit = f"brana-sched-{job_name}.timer"
    systemd_action = "start" if enabled else "stop"
    try:
        subprocess.run(
            ["systemctl", "--user", systemd_action, timer_unit],
            check=True, timeout=10,
        )
        console.print(f"  [{style}]Timer {systemd_action}ed.[/]\n")
    except subprocess.CalledProcessError:
        console.print(f"  [yellow]Warning: could not {systemd_action} timer {timer_unit}[/]\n")


@ops_app.command()
def sync(
    auto_commit: bool = typer.Option(False, "--auto-commit", help="Auto-commit changes"),
    direction: str = typer.Option("push", help="push or pull"),
):
    """Sync operational state (wraps sync-state.sh)."""
    root, _ = detect_project()
    if root is None:
        console.print("\n  Not in a git project.\n")
        raise typer.Exit(1)

    script = root / "system" / "scripts" / "sync-state.sh"
    if not script.exists():
        console.print(f"\n  sync-state.sh not found at {script}.\n")
        raise typer.Exit(1)

    cmd = ["bash", str(script), direction]
    if auto_commit:
        cmd.append("--auto-commit")

    console.print(f"\n  Running sync-state.sh {direction}...")
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=60, cwd=root)
        for line in result.stdout.strip().splitlines():
            console.print(f"  {line}")
        if result.returncode != 0:
            for line in result.stderr.strip().splitlines():
                console.print(f"  [red]{line}[/]")
            raise typer.Exit(1)
    except subprocess.TimeoutExpired:
        console.print("  [red]Timeout after 60s.[/]\n")
        raise typer.Exit(1)
    console.print()


@ops_app.command()
def reindex():
    """Reindex brana-knowledge into ruflo memory (wraps index-knowledge.sh)."""
    root, _ = detect_project()
    if root is None:
        console.print("\n  Not in a git project.\n")
        raise typer.Exit(1)

    script = root / "system" / "scripts" / "index-knowledge.sh"
    if not script.exists():
        console.print(f"\n  index-knowledge.sh not found at {script}.\n")
        raise typer.Exit(1)

    console.print("\n  Running index-knowledge.sh (this may take a while)...")
    try:
        result = subprocess.run(
            ["bash", str(script)], capture_output=True, text=True,
            timeout=600, cwd=root,
        )
        for line in result.stdout.strip().splitlines():
            console.print(f"  {line}")
        if result.returncode != 0:
            for line in result.stderr.strip().splitlines():
                console.print(f"  [red]{line}[/]")
            raise typer.Exit(1)
    except subprocess.TimeoutExpired:
        console.print("  [red]Timeout after 10m.[/]\n")
        raise typer.Exit(1)
    console.print()

File: system/cli/rust/src/commands/ops.rs
//! Ops subcommand handlers

use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::process::Command;

use crate::themes;
use crate::util::{find_project_root, home, load_scheduler, load_status};

// ── ops commands ────────────────────────────────────────────────────────

pub fn cmd_ops_status(theme: &themes::Theme) {
    let sched = load_scheduler();
    let status = load_status();
    let jobs = sched.get("jobs").and_then(|v| v.as_object()).cloned().unwrap_or_default();

    println!("\n{}Scheduler{}", themes::ansi(theme.color("header")), themes::RESET);
    for (name, cfg) in jobs.iter() {
        let enabled = cfg["enabled"].as_bool().unwrap_or(true);
        let job_st = status.get(name).and_then(|v| v["status"].as_str()).unwrap_or("—");
        let ts = status.get(name).and_then(|v| v["timestamp"].as_str()).unwrap_or("");
        let short_ts = if ts.len() > 16 { &ts[5..16] } else { ts };
        let schedule = cfg["schedule"].as_str().unwrap_or("—");
        let ic = if !enabled { theme.icon("parked") }
            else { match job_st { "SUCCESS" => theme.icon("done"), "FAILED"|"TIMEOUT" => theme.icon("blocked"), _ => theme.icon("pending") } };
        let col = if !enabled { themes::ansi("dim") }
            else { match job_st { "SUCCESS" => themes::ansi("green"), "FAILED"|"TIMEOUT" => themes::ansi("red"), _ => themes::ansi("yellow") } };
        let disabled = if !enabled { "  [disabled]" } else { "" };
        println!("{col}  {ic} {name:<24} {schedule:<24} {job_st:<10} {short_ts}{disabled}{}", themes::RESET);
    }
    println!();
}

pub fn cmd_ops_health(theme: &themes::Theme) {
    let sched = load_scheduler();
    let status = load_status();
    let jobs = sched.get("jobs").and_then(|v| v.as_object()).cloned().unwrap_or_default();

    let mut failures = vec![];
    let mut skipped = vec![];
    for (name, info) in &status {
        if let Some(st) = info["status"].as_str() {
            match st {
                "FAILED" => failures.push(name.as_str()),
                "SKIPPED" => skipped.push(name.as_str()),
                _ => {}
            }
        }
    }

    let collisions = find_collisions(&jobs);

    println!("\n{}Scheduler health{}", themes::ansi(theme.color("header")), themes::RESET);
    if failures.is_empty() {
        println!("{}  {} No failures{}", themes::ansi("green"), theme.icon("done"), themes::RESET);
    } else {
        println!("{}  {} Failures: {}{}", themes::ansi("red"), theme.icon("blocked"), failures.join(", "), themes::RESET);
    }
    if !skipped.is_empty() {
        println!("{}  {} Skipped: {}{}", themes::ansi("yellow"), theme.icon("pending"), skipped.join(", "), themes::RESET);
    }
    if collisions.is_empty() {
        println!("{}  {} No schedule collisions{}", themes::ansi("green"), theme.icon("done"), themes::RESET);
    } else {
        println!("{}  {} Schedule collisions:{}", themes::ansi("red"), theme.icon("blocked"), themes::RESET);
        for (sched, proj, names) in &collisions {
            println!("      {sched} on {proj}: {}", names.join(", "));
        }
    }
    let enabled = jobs.values().filter(|v| v["enabled"].as_bool().unwrap_or(true)).count();
    println!("\n  {enabled} enabled, {} disabled, {} total\n", jobs.len() - enabled, jobs.len());
}

pub fn cmd_ops_collisions(theme: &themes::Theme) {
    let sched = load_scheduler();
    let jobs = sched.get("jobs").and_then(|v| v.as_object()).cloned().unwrap_or_default();
    let collisions = find_collisions(&jobs);
    if collisions.is_empty() {
        println!("\n  {}{}  No schedule collisions.{}\n", themes::ansi("green"), theme.icon("done"), themes::RESET);
    } else {
        println!("\n{}Schedule collisions{}", themes::ansi(theme.color("header")), themes::RESET);
        for (sched, proj, names) in &collisions {
            println!("{}  {} {sched} on {proj}: {}{}", themes::ansi("red"), theme.icon("blocked"), names.join(", "), themes::RESET);
        }
        println!();
    }
}

pub fn cmd_ops_drift(theme: &themes::Theme) {
    let root = find_project_root().unwrap_or_else(|| { eprintln!("Not in git repo"); std::process::exit(1); });
    let template_path = root.join("system/scheduler/scheduler.template.json");
    let live = load_scheduler();
    let template: HashMap<String, serde_json::Value> = std::fs::read_to_string(&template_path)
        .ok()
        .and_then(|c| serde_json::from_str(&c).ok())
        .unwrap_or_default();

    let tmpl_jobs = template.get("jobs").and_then(|v| v.as_object()).cloned().unwrap_or_default();
    let live_jobs = live.get("jobs").and_then(|v| v.as_object()).cloned().unwrap_or_default();

    let mut drifts = vec![];
    let all_names: std::collections::BTreeSet<_> = tmpl_jobs.keys().chain(live_jobs.keys()).collect();
    for name in all_names {
        if !tmpl_jobs.contains_key(name) {
            drifts.push(format!("{}+ {name}: in live but not in template{}", themes::ansi("yellow"), themes::RESET));
        } else if !live_jobs.contains_key(name) {
            drifts.push(format!("{}- {name}: in template but not in live{}", themes::ansi("red"), themes::RESET));
        } else {
            for field in &["schedule", "enabled", "command", "project", "type"] {
                let tv = &tmpl_jobs[name][field];
                let lv = &live_jobs[name][field];
                if tv != lv {
                    drifts.push(format!("{}~ {name}.{field}: template={tv} live={lv}{}", themes::ansi("yellow"), themes::RESET));
                }
            }
        }
    }

    if drifts.is_empty() {
        println!("\n  {}{}  No drift — live matches template.{}\n", themes::ansi("green"), theme.icon("done"), themes::RESET);
    } else {
        println!("\n{}Config drift (template vs live){}", themes::ansi(theme.color("header")), themes::RESET);
        for d in &drifts { println!("  {d}"); }
        println!();
    }
}

pub fn cmd_ops_logs(job_name: &str, tail: usize) {
    let log_dir = home().join(format!(".claude/scheduler/logs/{job_name}"));
    if !log_dir.exists() {
        eprintln!("No logs for '{job_name}'."); std::process::exit(1);
    }
    let mut logs: Vec<_> = std::fs::read_dir(&log_dir).unwrap()
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().map_or(false, |x| x == "log"))
        .collect();
    logs.sort_by_key(|e| std::cmp::Reverse(e.file_name()));
    if let Some(latest) = logs.first() {
        let content = std::fs::read_to_string(latest.path()).unwrap_or_default();
        let lines: Vec<_> = content.lines().collect();
        let start = if lines.len() > tail { lines.len() - tail } else { 0 };
        println!("\n  \x1b[2m{}\x1b[0m\n", latest.file_name().to_string_lossy());
        for line in &lines[start..] {
            if line.contains("SUCCESS") { println!("  \x1b[32m{line}\x1b[0m"); }
            else if line.contains("FAILED") || line.contains("ERROR") { println!("  \x1b[31m{line}\x1b[0m"); }
            else { println!("  {line}"); }
        }
        println!();
    }
}

pub fn cmd_ops_history(job_name: &str, last: usize, theme: &themes::Theme) {
    let log_dir = home().join(format!(".claude/scheduler/logs/{job_name}"));
    if !log_dir.exists() { eprintln!("No history for '{job_name}'."); std::process::exit(1); }
    let mut logs: Vec<_> = std::fs::read_dir(&log_dir).unwrap()
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().map_or(false, |x| x == "log"))
        .collect();
    logs.sort_by_key(|e| std::cmp::Reverse(e.file_name()));

    println!("\n{}History: {job_name} (last {}){}", themes::ansi(theme.color("header")), logs.len().min(last), themes::RESET);
    for entry in logs.iter().take(last) {
        let content = std::fs::read_to_string(entry.path()).unwrap_or_default();
        let (st, col) = if content.contains("SUCCESS") { ("SUCCESS", "green") }
            else if content.contains("FAILED") { ("FAILED", "red") }
            else if content.contains("TIMEOUT") { ("TIMEOUT", "red") }
            else if content.contains("SKIPPED") { ("SKIPPED", "dim") }
            else { ("UNKNOWN", "yellow") };
        let ic = match st { "SUCCESS" => theme.icon("done"), "FAILED"|"TIMEOUT" => theme.icon("blocked"), _ => theme.icon("pending") };
        let date = entry.path().file_stem().map(|s| s.to_string_lossy().to_string()).unwrap_or_default();
        println!("{}{ic} {date}  {st}{}", themes::ansi(col), themes::RESET);
    }
    println!();
}

pub fn cmd_ops_run(job_name: &str) {
    validate_job_name(job_name);
    let unit = format!("brana-sched-{job_name}.service");
    println!("\n  Starting {unit}...");
    let status = Command::new("systemctl").args(["--user", "start", &unit]).status();
    match status {
        Ok(s) if s.success() => println!("  \x1b[32mTriggered. Check: brana ops logs {job_name}\x1b[0m\n"),
        _ => { eprintln!("  \x1b[31mFailed to start {unit}\x1b[0m"); std::process::exit(1); }
    }
}

pub fn cmd_ops_toggle(job_name: &str, enabled: bool) {
    validate_job_name(job_name);
    let config_path = home().join(".claude/scheduler/scheduler.json");
    let mut sched = load_scheduler();
    if let Some(jobs) = sched.get_mut("jobs").and_then(|v| v.as_object_mut()) {
        if let Some(job) = jobs.get_mut(job_name) {
            job["enabled"] = serde_json::Value::Bool(enabled);
        } else {
            eprintln!("Job '{job_name}' not found."); std::process::exit(1);
        }
    }
    std::fs::write(&config_path, serde_json::to_string_pretty(&sched).unwrap() + "\n").ok();
    let action = if enabled { "Enabled" } else { "Disabled" };
    let col = if enabled { "\x1b[32m" } else { "\x1b[33m" };
    println!("\n  {col}{action} '{job_name}'{}", themes::RESET);

    let timer = format!("brana-sched-{job_name}.timer");
    let cmd = if enabled { "start" } else { "stop" };
    if Command::new("systemctl").args(["--user", cmd, &timer]).status().is_ok() {
        println!("  {col}Timer {cmd}ed.{}\n", themes::RESET);
    }
}

pub fn cmd_ops_sync(direction: &str, auto_commit: bool) {
    let root = find_project_root().unwrap_or_else(|| { eprintln!("Not in git repo"); std::process::exit(1); });
    let script = root.join("system/scripts/sync-state.sh");
    if !script.exists() { eprintln!("sync-state.sh not found"); std::process::exit(1); }

    let mut cmd = Command::new("bash");
    cmd.arg(&script).arg(direction).current_dir(&root);
    if auto_commit { cmd.arg("--auto-commit"); }

    println!("\n  Running sync-state.sh {direction}...");
    match cmd.status() {
        Ok(s) if s.success() => println!("  \x1b[32mDone.\x1b[0m\n"),
        _ => { eprintln!("  \x1b[31mFailed.\x1b[0m"); std::process::exit(1); }
    }
}

pub fn cmd_ops_reindex() {
    let root = find_project_root().unwrap_or_else(|| { eprintln!("Not in git repo"); std::process::exit(1); });
    let script = root.join("system/scripts/index-knowledge.sh");
    if !script.exists() { eprintln!("index-knowledge.sh not found"); std::process::exit(1); }

    println!("\n  Running index-knowledge.sh...");
    match Command::new("bash").arg(&script).current_dir(&root).status() {
        Ok(s) if s.success() => println!("  \x1b[32mDone.\x1b[0m\n"),
        _ => { eprintln!("  \x1b[31mFailed.\x1b[0m"); std::process::exit(1); }
    }
}

// ── ops metrics ──────────────────────────────────────────────────────────

pub fn cmd_ops_metrics(session_file: &PathBuf) {
    let content = std::fs::read_to_string(session_file).unwrap_or_default();
    let events: Vec<serde_json::Value> = content
        .lines()
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect();

    let total = events.len();
    let successes = events.iter().filter(|e| e["outcome"].as_str() == Some("success")).count();
    let failures = events.iter().filter(|e| matches!(e["outcome"].as_str(), Some("failure" | "test-fail" | "lint-fail"))).count();
    let corrections = events.iter().filter(|e| e["outcome"].as_str() == Some("correction")).count();
    let test_writes = events.iter().filter(|e| e["outcome"].as_str() == Some("test-write")).count();
    let cascades = events.iter().filter(|e| e["cascade"].as_bool() == Some(true)).count();
    let pr_creates = events.iter().filter(|e| e["outcome"].as_str() == Some("pr-create")).count();
    let test_passes = events.iter().filter(|e| e["outcome"].as_str() == Some("test-pass")).count();
    let test_fails = events.iter().filter(|e| e["outcome"].as_str() == Some("test-fail")).count();
    let lint_passes = events.iter().filter(|e| e["outcome"].as_str() == Some("lint-pass")).count();
    let lint_fails = events.iter().filter(|e| e["outcome"].as_str() == Some("lint-fail")).count();
    let edits = events.iter().filter(|e| matches!(e["tool"].as_str(), Some("Edit" | "Write"))).count();
    let delegations = events.iter().filter(|e| e["tool"].as_str() == Some("Task")).count();

    let tools: std::collections::BTreeSet<&str> = events.iter().filter_map(|e| e["tool"].as_str()).collect();
    let files: std::collections::BTreeSet<&str> = events.iter()
        .filter_map(|e| e["detail"].as_str())
        .filter(|d| !d.is_empty())
        .collect();
    let files_vec: Vec<&str> = files.into_iter().take(10).collect();

    let correction_rate = if edits > 0 { corrections as f64 / edits as f64 } else { 0.0 };
    let test_write_rate = if edits > 0 { test_writes as f64 / edits as f64 } else { 0.0 };
    let cascade_rate = if failures > 0 { cascades as f64 / failures as f64 } else { 0.0 };

    let mut fail_files: HashSet<String> = HashSet::new();
    let mut auto_fixes = 0usize;
    for e in &events {
        let detail = e["detail"].as_str().unwrap_or("").to_string();
        match e["outcome"].as_str() {
            Some("failure" | "test-fail" | "lint-fail") => { fail_files.insert(detail); }
            Some("success" | "correction" | "test-pass" | "lint-pass") => {
                if fail_files.remove(&detail) { auto_fixes += 1; }
            }
            _ => {}
        }
    }
    let auto_fix_rate = if failures > 0 { auto_fixes as f64 / failures as f64 } else { 0.0 };

    let test_total = test_passes + test_fails;
    let test_pass_rate = if test_total > 0 { format!("{:.2}", test_passes as f64 / test_total as f64) } else { "N/A".into() };
    let lint_total = lint_passes + lint_fails;
    let lint_pass_rate = if lint_total > 0 { format!("{:.2}", lint_passes as f64 / lint_total as f64) } else { "N/A".into() };

    let output = serde_json::json!({
        "events": total,
        "successes": successes,
        "failures": failures,
        "corrections": corrections,
        "test_writes": test_writes,
        "cascades": cascades,
        "pr_creates": pr_creates,
        "edits": edits,
        "test_passes": test_passes,
        "test_fails": test_fails,
        "lint_passes": lint_passes,
        "lint_fails": lint_fails,
        "delegations": delegations,
        "flywheel": {
            "correction_rate": format!("{:.2}", correction_rate),
            "auto_fix_rate": format!("{:.2}", auto_fix_rate),
            "test_write_rate": format!("{:.2}", test_write_rate),
            "cascade_rate": format!("{:.2}", cascade_rate),
            "test_pass_rate": test_pass_rate,
            "lint_pass_rate": lint_pass_rate,
            "delegations": delegations,
            "pr_creates": pr_creates,
        },
        "tools": tools.into_iter().collect::<Vec<_>>().join(","),
        "files": files_vec.join(","),
    });

    println!("{}", serde_json::to_string(&output).unwrap());
}

// ── helpers ─────────────────────────────────────────────────────────────

fn find_collisions(jobs: &serde_json::Map<String, serde_json::Value>) -> Vec<(String, String, Vec<String>)> {
    let mut groups: HashMap<(String, String), Vec<String>> = HashMap::new();
    for (name, cfg) in jobs {
        if !cfg["enabled"].as_bool().unwrap_or(true) { continue; }
        let key = (
            cfg["schedule"].as_str().unwrap_or("").to_string(),
            std::path::Path::new(cfg["project"].as_str().unwrap_or(""))
                .file_name().unwrap_or_default().to_string_lossy().to_string(),
        );
        groups.entry(key).or_default().push(name.clone());
    }
    groups.into_iter()
        .filter(|(_, v)| v.len() > 1)
        .map(|((s, p), v)| (s, p, v))
        .collect()
}

fn validate_job_name(name: &str) {
    if !name.chars().all(|c| c.is_alphanumeric() || c == '-' || c == '_') {
        eprintln!("Invalid job name '{name}'. Use alphanumeric, hyphens, underscores.");
        std::process::exit(1);
    }
}
